internal/controller/openstacknodeimagerelease_controller.go

```diff
@@ -46,18 +69,274 @@ type OpenStackNodeImageReleaseReconciler struct {
+
+	case images.ImageStatusDeactivated, images.ImageStatusKilled:
+		// These statuses are unexpected. Hence we set a failure for them. See the explanation below:
+		// `deactivated`: image is not allowed to use to any non-admin user
+		// `killed`: an error occurred during the uploading of an image’s data, and that the image is not readable (via v1 API)
+		err = fmt.Errorf("image status %s is unexpected", image.Status)
+		conditions.MarkFalse(openstacknodeimagerelease,
+			apiv1alpha1.OpenStackImageReadyCondition,
+			apiv1alpha1.IssueWithOpenStackImageReason,
+			clusterv1beta1.ConditionSeverityError,
+			err.Error(),
+		)
+		openstacknodeimagerelease.Status.Ready = false
+		return ctrl.Result{}, err
+
+	case images.ImageStatusQueued, images.ImageStatusSaving, images.ImageStatusDeleted, images.ImageStatusPendingDelete, images.ImageStatusImporting:
+		// The other statuses are expected. See the explanation below:
+		// - `deleted`, `pending_delete`: The image has been deleted and will be removed soon, hence we can create and import the image in the next reconciliation loop.
+		// - `importing`, `uploading`, `saving`, and `queued`: The image is in the process of being uploaded or imported via upload to Glance or the Glance image import API (performed by this reconciliation loop). Therefore, let's wait for it.
+		logger.Info("OpenStackNodeImageRelease **not ready** yet - waiting for image to become ACTIVE", "name", openstacknodeimagerelease.Spec.Image.CreateOpts.Name, "ID", imageID, "status", image.Status)
+		conditions.MarkFalse(openstacknodeimagerelease, apiv1alpha1.OpenStackImageReadyCondition, apiv1alpha1.OpenStackImageNotImportedYetReason, clusterv1beta1.ConditionSeverityInfo, "waiting for image to become ACTIVE")
+		openstacknodeimagerelease.Status.Ready = false
+		// Wait for image
+		return ctrl.Result{RequeueAfter: waitForImageBecomeActive}, nil
+
+	default:
+		// An unknown state - set failure
+		err = fmt.Errorf("image status %s is unknown", image.Status)
+		conditions.MarkFalse(openstacknodeimagerelease,
+			apiv1alpha1.OpenStackImageReadyCondition,
+			apiv1alpha1.IssueWithOpenStackImageReason,
+			clusterv1beta1.ConditionSeverityError,
+			err.Error(),
+		)
+		openstacknodeimagerelease.Status.Ready = false
+		return ctrl.Result{}, err
+	}
+
+	// Requeue to ensure the image's presence
+	return ctrl.Result{Requeue: true, RequeueAfter: reconcileImage}, nil
+}
+
+func (r *OpenStackNodeImageReleaseReconciler) getCloudFromSecret(ctx context.Context, secretNamespace, secretName, cloudName string) (clientconfig.Cloud, error) {
+	var clouds clientconfig.Clouds
+	emptyCloud := clientconfig.Cloud{}
+
+	if cloudName == "" {
+		return emptyCloud, fmt.Errorf("secret name set to %s but no cloud was specified. Please set cloud_name in your machine spec", secretName)
+	}
+
+	secret := &corev1.Secret{}
+	err := r.Get(ctx, types.NamespacedName{
+		Namespace: secretNamespace,
+		Name:      secretName,
+	}, secret)
+	if err != nil {
```

Comment:
Please see 597fc5d7fb725d0e1a2b3ece10e595752949bbfc

Reply:
I actually think that we should instead have a condition for this and set it on false instead of returning the error.. Otherwise, we will have a loop where the controller always tries to find the secret which does not exist. 
Instead, the user should have a meaningful error message in the conditions which show him that he forgot to apply the secret

Reply:
Thank you for your comment. Now the user can differentiate between errors. I did it in 7d7345b5e0607c44b8d5badb32243b19cde349c0.

Reply:
perfect :)